lib/attr_extras/attr_initialize.rb

```diff
@@ -14,20 +16,21 @@ def apply
     validate_arity = method(:validate_arity)
     set_ivar_from_hash = method(:set_ivar_from_hash)
 
+    declaration = AttributesDeclaration.new(names)
+
     klass.send(:define_method, :initialize) do |*values|
       validate_arity.call(values.length, self.class)
 
-      names.zip(values).each do |name_or_names, value|
-        if name_or_names.is_a?(Array)
-          hash = value || {}
-
-          name_or_names.each do |name|
-            set_ivar_from_hash.call(self, name, hash)
-          end
-        else
-          name = name_or_names
-          instance_variable_set("@#{name}", value)
-        end
+      positional_attributes_values = values.take(declaration.positional_attributes.count)
+      hash_attributes_values = values.drop(declaration.positional_attributes.count)
```

Comment:
I wonder if a more tell-don't-asky version of this would read well: maybe initializing the declaration with the values, or passing the values into an instance method. `declaration.positional_values` or `declaration.positional_values(values)` What do you think?